Hardware/Hardware/CardReader/CardReaderServiceError.swift

```diff
@@ -1,20 +1,152 @@
 /// Models errors thrown by the CardReaderService.
-/// This is doing the bare minimum for now.
-/// Proper error handling is coming in
-/// https://github.com/woocommerce/woocommerce-ios/issues/3734
+/// It identifies the interaction with the card reader
+/// where the error was thrown.
+/// It is important to mark the specific interaction
+/// because some operations (like, for example, procesing a payment)
+/// require of three interactions with the SDK in sequence, and any error
+/// in any of the steps in that sequence would make the whole operation fail
 public enum CardReaderServiceError: Error {
     /// Error thrown during reader discovery
-    case discovery
+    case discovery(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while connecting to a reader
-    case connection
+    case connection(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while creating a payment intent
-    case intentCreation
+    case intentCreation(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while collecting payment methods
-    case paymentMethod
+    case paymentMethod(underlyingError: UnderlyingError = .internalServiceError)
 
     /// Error thrown while capturing a payment
-    case capturePayment
+    case capturePayment(underlyingError: UnderlyingError = .internalServiceError)
+}
+
+
+/// Underlying error. Models the specific error that made a given
+/// interaction with the SDK fail.
+public enum UnderlyingError: Error {
+    /// The service is busy executing another command. The service can only execute a single command at a time.
+    case busy
```

Comment:
So, some of these errors are because we coded something wrong, e.g. busy, or because the user did something wrong, e.g. left a card in the reader from an earlier transaction.

Instead of having the caller have to figure out what to do in each case, what if the enumeration provided a localized user facing strings for cases where there is something the user could do, like a title "Card Left in Reader" and a suggested action "Please remove your card and try again"

If there is nothing the user could do, we could use the title to bubble up a unique code or something to the end user so they can tell it to us when reporting the error, e.g. "An Unexpected Error Occurred (10223)" "Please try connecting to the card reader again"

Reply:
Right. I logged a separate issue for adding user-facing strings to these underlying errors: #3868 

The idea would be that UnderlyingError provides a description that could be used as default for that specific error. 

Later on, the UI could "override" that specific error description if needed.

Reply:
WFM. Thank you!